keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,428 @@
+
+Similar features for protecting PV and PVC already exist as [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md).
+
+
+## Motivation
+
+This feature aims to protect secrets from deleting while they are in-use.
+Secrets can be used by below ways:
+- From Pod:
+  - [Mounted as data volumes](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables)
+  - [Generic ephemeral volumes
+](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes) (can be handled as CSI PV below)
+- From PV:
+  - [CSI](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html):
+    - provisioner secret
+    - controller publish secret
+    - node stage secret
+    - node publish secret
+    - controller expand secret
+  - non-CSI:
+    - dependent on each storage driver and will be deprecated soon (Out of scope)
+- [From Snapshot](https://kubernetes-csi.github.io/docs/secrets-and-credentials-volume-snapshot-class.html):
+  - snapshotter secret
+
+### Goals
+
+- Protect secrets from deleting while they are in use.
+
+### Non-Goals
+
+- Protect important secrets that aren't in use from deleting
+- Protect other resources than secret from deleting.
+
+## Proposal
+
+A new controller to protect secret is introduced.
+
+### User Stories
+
+#### Story 1
+
+A user creates a secret and a pod using the secret. Then, the user mistakenly delete the secret while the pod is using it.
+The secret is protected until the pod using the secret is deleted.
+
+#### Story 2
+
+A user creates a volume that uses a certain secret in the same namespace. Then, the user delete the namespace.
+The secret is protected until the volume using the secret is deleted and the deletion of the volume succeeds.
+
+#### Story 3
+
+A user really would like to delete a secret inspite that it is used by other resources.
+The user force delete the secret while it is used by other resources, and the secret isn't protected and is actually deleted.
+
+### Notes/Constraints/Caveats (Optional)
+
+- Compatibility:
+  - There might be many existing scripts that don't care the order of deletion. Therefore, such scripts might stuck on secret deletion, if the deletion of the resources using the secrets are done later.
+- Usability:
+  - Use of the secret in other resource will not be obvious to users. Therefore, users might not easily understand why the secret is not deleted.
+  - Users might need to force delete the secret on deletion or would like to avoid protection for certain secrets that already exist or that are newly created.
+
+### Risks and Mitigations
+
```

Comment:
Adding as a comment here, for not sure where to add this kind of information.

There will be a corner case when Reclaim Policy is `Retain` and PV is referencing secrets in a certain user  namespace. In such a case, when PVC is deleted, the PV referencing secrets remains as a resource not managed by the namespace, as a result, deletion of the secrets is blocked. Howerver, the user of the namespace won't notice why it is blocked and won't find a way to delete the secrets. It should be a right behavior in terms of protecting in-use secret, but it isn't user friendly.

In addition, to make things worse, provisioner secrets that are defined by using template with PVC information, like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret, because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted. This behavior is inconsistent with other secrets, like node publish secret which is stored in PV.

Reply:
What should user do in this situation?  Use force delete?

Can you add this to the doc?

Reply:
Added this corner case here and workaround in Design Details section.